Stackage/HEAD/History.hs

```diff
@@ -1,90 +1,172 @@
 import Data.Bifunctor (second)
-import Data.Hashable (Hashable)
 import Data.List.NonEmpty (NonEmpty (..))
-import Data.Maybe (fromMaybe, fromJust)
+import Data.Maybe (fromJust)
 import Data.Semigroup ((<>))
 import Data.Text (Text)
+import Data.Typeable (Typeable)
 import Data.Vector (Vector)
-import Data.Void
+import GHC.Generics
+import Path
 import Path.IO (forgivingAbsence)
-import System.FilePath (replaceExtension)
-import Text.Megaparsec
-import Text.Megaparsec.Char (digitChar, hexDigitChar, string)
 import Text.URI (URI)
 import qualified Data.ByteString.Lazy as BL
 import qualified Data.Csv             as Csv
 import qualified Data.Text            as T
-import qualified Data.Text.Encoding   as TE
 import qualified Data.Vector          as V
 import qualified Text.URI             as URI
 import qualified Text.URI.QQ          as Q
 
--- | Build history.
-
-newtype History = History (Vector HistoryItem)
+----------------------------------------------------------------------------
+-- History item
 
 -- | Build history item.
 
-newtype HistoryItem = HistoryItem
-  { unHistoryItem :: Text
-  } deriving (Eq, Hashable)
+data HistoryItem = HistoryItem
+  { historyItemTarget :: !Text -- ^ Target name
+  , historyItemSha1 :: !Text -- ^ Commit SHA1
+  , historyItemBuildUrl :: !URI -- ^ Link to the build
+  } deriving (Eq, Ord, Generic)
 
 instance Csv.FromRecord HistoryItem where
   parseRecord v = do
-    bs <- v Csv..! 0
-    (return . HistoryItem . TE.decodeUtf8) bs
+    let fromMonadThrow = either (fail . displayException) return
+    target   <- v Csv..! 0
+    commit   <- v Csv..! 1
+    buildUrl <- (v Csv..! 2) >>= fromMonadThrow . URI.mkURI
+    fromMonadThrow (mkHistoryItem target commit buildUrl)
 
 instance Csv.ToRecord HistoryItem where
-  toRecord (HistoryItem txt) =
-    Csv.record [TE.encodeUtf8 txt]
+  toRecord HistoryItem {..} = Csv.record
+    [ Csv.toField historyItemTarget
+    , Csv.toField historyItemSha1
+    , Csv.toField (URI.render historyItemBuildUrl)
+    ]
+
+-- | Exception that is thrown when 'HistoryItem' cannot be constructed.
+
+data HistoryItemException
+  = InvalidHistoryItem !Text !Text !URI
+  deriving (Eq, Ord, Show, Typeable)
+
+instance Exception HistoryItemException
+
+-- | Create a 'HistoryItem' value.
+
+mkHistoryItem :: MonadThrow m
+  => Text              -- ^ Target name
+  -> Text              -- ^ Commit SHA1
+  -> URI               -- ^ Link to the build
+  -> m HistoryItem
+mkHistoryItem target sha1 buildUrl =
+  maybe (throwM $ InvalidHistoryItem target sha1 buildUrl) return $ do
+    let item = HistoryItem target sha1 buildUrl
+    -- Assert that reportPath is obtainable
+    void (reportPathMaybe item)
+    -- Assert that logPath is obtainable
+    void (logsPathMaybe item)
+    -- Assert that target is a valid RText 'PathPiece
+    void (URI.mkPathPiece target)
+    -- Assert that commit is a valid RText 'PathPiece
+    void (URI.mkPathPiece sha1)
+    return item
```

Comment:
The preconditions should be made explicit in the documentation of the smart constructor.